### PR TITLE
persist: use a single main for all benches

### PR DESCRIPTION
**Pull request by danhhz**
There's no particular reason they need to be separate binaries and only
having one cuts down on link time. While I'm in here, also:

- Fix `cargo bench -p persist -- --baseline=foo` (broken previously
  because lib was getting the std bench harness for its non-existent
  benches and the harness errors on unknown flags).
- Standardize on passing a criterion::BenchmarkGroup instead of
  criterion::Criterion to "top-level" benchmarks.
- Skip the reasonably expensive setup work in end_to_end when it wasn't
  selected for execution.
- Use the same DataGenerator for all benchmarks.
- Silence the criterion "html_reports" warning.
- Rename the benchmarks a bit to remove redundances in the names.

New output is much cleaner

    $ cargo bench -p persist
       Compiling persist v0.0.0
        Finished bench [optimized + debuginfo] target(s) in 32.28s
         Running unittests (target/release/deps/main-0ce35e00c69ce2e6)
    Gnuplot not found, using plotters backend
    MZ_PERSIST_RECORD_COUNT=819200 MZ_PERSIST_RECORD_SIZE_BYTES=64 MZ_PERSIST_BATCH_MAX_COUNT=131072
    Benchmarking end_to_end/replay/50MiB: Warming up for 3.0000 s
    Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 41.3s, or reduce sample count to 10.
    end_to_end/replay/50MiB time:   [405.32 ms 406.61 ms 407.95 ms]
                            thrpt:  [122.56 MiB/s 122.97 MiB/s 123.36 MiB/s]
    Found 4 outliers among 100 measurements (4.00%)
      4 (4.00%) high mild

    Benchmarking snapshot/mem/unsealed/50MiB: Warming up for 3.0000 s
    Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 9.3s, or reduce sample count to 50.
    snapshot/mem/unsealed/50MiB
                            time:   [91.837 ms 92.120 ms 92.406 ms]
                            thrpt:  [541.09 MiB/s 542.77 MiB/s 544.44 MiB/s]
    Found 4 outliers among 100 measurements (4.00%)
      2 (2.00%) low mild
      2 (2.00%) high mild
    Benchmarking snapshot/mem/trace/50MiB: Warming up for 3.0000 s
    Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 10.0s, or reduce sample count to 50.
    snapshot/mem/trace/50MiB
                            time:   [93.718 ms 93.964 ms 94.236 ms]
                            thrpt:  [530.58 MiB/s 532.12 MiB/s 533.52 MiB/s]
    Found 4 outliers among 100 measurements (4.00%)
      3 (3.00%) high mild
      1 (1.00%) high severe

    Benchmarking snapshot/file/unsealed/50MiB: Warming up for 3.0000 s
    Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 10.3s, or reduce sample count to 40.
    snapshot/file/unsealed/50MiB
                            time:   [100.44 ms 100.85 ms 101.29 ms]
                            thrpt:  [493.64 MiB/s 495.77 MiB/s 497.80 MiB/s]
    Found 9 outliers among 100 measurements (9.00%)
      1 (1.00%) low severe
      1 (1.00%) low mild
      5 (5.00%) high mild
      2 (2.00%) high severe
    Benchmarking snapshot/file/trace/50MiB: Warming up for 3.0000 s
    Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 10.6s, or reduce sample count to 40.
    snapshot/file/trace/50MiB
                            time:   [93.832 ms 94.086 ms 94.368 ms]
                            thrpt:  [529.84 MiB/s 531.43 MiB/s 532.87 MiB/s]
    Found 7 outliers among 100 measurements (7.00%)
      5 (5.00%) high mild
      2 (2.00%) high severe

    writer/log/mem_sync     time:   [39.419 ns 40.356 ns 41.133 ns]
    Found 2 outliers among 100 measurements (2.00%)
      2 (2.00%) low mild
    writer/log/file_sync    time:   [18.324 ms 18.551 ms 18.823 ms]
    Found 6 outliers among 100 measurements (6.00%)
      6 (6.00%) high severe

    writer/log/mem/50MiB    time:   [6.5780 ms 6.6923 ms 6.8125 ms]
                            thrpt:  [7.1675 GiB/s 7.2962 GiB/s 7.4230 GiB/s]
    Found 1 outliers among 100 measurements (1.00%)
      1 (1.00%) high severe
    Benchmarking writer/log/file/50MiB: Warming up for 1.0000 s
    Warning: Unable to complete 100 samples in 1.0s. You may wish to increase target time to 4.3s, or reduce sample count to 20.
    writer/log/file/50MiB   time:   [41.201 ms 41.862 ms 42.560 ms]
                            thrpt:  [1.1473 GiB/s 1.1664 GiB/s 1.1851 GiB/s]
    Found 15 outliers among 100 measurements (15.00%)
      2 (2.00%) low severe
      2 (2.00%) low mild
      8 (8.00%) high mild
      3 (3.00%) high severe

    writer/encode_batch/unsealed/50MiB
                            time:   [19.726 ms 20.059 ms 20.402 ms]
                            thrpt:  [2.3933 GiB/s 2.4342 GiB/s 2.4753 GiB/s]
    writer/encode_batch/trace/50MiB
                            time:   [19.564 ms 19.819 ms 20.092 ms]
                            thrpt:  [2.4302 GiB/s 2.4637 GiB/s 2.4959 GiB/s]
    Found 13 outliers among 100 measurements (13.00%)
      13 (13.00%) high mild

    writer/blob_cache_set_unsealed_batch/file_unsorted/50MiB
                            time:   [63.956 ms 64.511 ms 65.203 ms]
                            thrpt:  [766.83 MiB/s 775.06 MiB/s 781.79 MiB/s]
    Found 1 outliers among 10 measurements (10.00%)
      1 (10.00%) high mild
    Benchmarking writer/blob_cache_set_unsealed_batch/mem_unsorted/50MiB: Warming up for 1.0000 s
    Warning: Unable to complete 10 samples in 1.0s. You may wish to increase target time to 1.4s or enable flat sampling.
    writer/blob_cache_set_unsealed_batch/mem_unsorted/50MiB
                            time:   [32.449 ms 32.930 ms 33.750 ms]
                            thrpt:  [1.4468 GiB/s 1.4828 GiB/s 1.5048 GiB/s]
    Found 2 outliers among 10 measurements (20.00%)
      1 (10.00%) low severe
      1 (10.00%) high mild

    Benchmarking writer/indexed_drain/mem_sorted/50MiB: Warming up for 3.0000 s
    Warning: Unable to complete 10 samples in 5.0s. You may wish to increase target time to 5.9s or enable flat sampling.
    writer/indexed_drain/mem_sorted/50MiB
                            time:   [91.524 ms 109.44 ms 117.71 ms]
                            thrpt:  [424.79 MiB/s 456.89 MiB/s 546.30 MiB/s]
    Benchmarking writer/indexed_drain/mem_unsorted/50MiB: Warming up for 3.0000 s
    Warning: Unable to complete 10 samples in 5.0s. You may wish to increase target time to 7.9s or enable flat sampling.
    writer/indexed_drain/mem_unsorted/50MiB
                            time:   [105.28 ms 115.92 ms 120.74 ms]
                            thrpt:  [414.11 MiB/s 431.34 MiB/s 474.93 MiB/s]
    Benchmarking writer/indexed_drain/file_sorted/50MiB: Warming up for 3.0000 s
    Warning: Unable to complete 10 samples in 5.0s. You may wish to increase target time to 8.6s or enable flat sampling.
    writer/indexed_drain/file_sorted/50MiB
                            time:   [143.87 ms 144.90 ms 146.49 ms]
                            thrpt:  [341.32 MiB/s 345.06 MiB/s 347.53 MiB/s]
    Found 3 outliers among 10 measurements (30.00%)
      1 (10.00%) low mild
      1 (10.00%) high mild
      1 (10.00%) high severe
    writer/indexed_drain/file_unsorted/50MiB
                            time:   [148.95 ms 151.90 ms 154.85 ms]
                            thrpt:  [322.89 MiB/s 329.16 MiB/s 335.67 MiB/s]

Cleaning up all the "Warning: Unable to complete 10 samples" messages is
left as a followup.

### Motivation

   * This PR refactors existing code.

### Tips for reviewer

I'm about to add s3 variants of some of our benchmarks and the organic structure was bugging me :).

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [N/A] This PR adds a release note for any [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

### CODE REVIEW

File: src/persist/benches/benches.rs [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+pub fn bench_persist(c: &mut Criterion) {
+    let data = DataGenerator::default();
+
+    end_to_end::bench_replay(&data, &mut c.benchmark_group("end_to_end"));
+
+    snapshot::bench_mem(&data, &mut c.benchmark_group("snapshot/mem"));
+    snapshot::bench_file(&data, &mut c.benchmark_group("snapshot/file"));
+
+    writer::bench_log(&mut c.benchmark_group("writer/log"));
+    writer::bench_blob(&data, &mut c.benchmark_group("writer/log"));
```
Comment (danhhz) [inline src/persist/benches/benches.rs:26]: note to self: s/log/blob/

File: src/persist/benches/benches.rs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+use persist::workload::DataGenerator;
+
+mod end_to_end;
+mod snapshot;
+mod writer;
+
+pub fn bench_persist(c: &mut Criterion) {
+    let data = DataGenerator::default();
+
+    end_to_end::bench_replay(&data, &mut c.benchmark_group("end_to_end"));
```
Comment (ruchirK) [inline src/persist/benches/benches.rs:20]: very very minor nit: should we follow the rough convention in the rest of the function and name this "end_to_end/replay"?

**Review by ruchirK — APPROVED** [review]
I skimmed everything twice - LGTM!

File: src/persist/benches/benches.rs (hunk 2 above)
Comment (danhhz) [reply src/persist/benches/benches.rs]: yeah, good call

File: src/persist/benches/benches.rs (hunk 1 above)
Comment (danhhz) [reply src/persist/benches/benches.rs]: done

**Review by danhhz — COMMENTED** [review]
TFTR!

**Pull request merged**